README.md
Data List View: Data Items List, Data Actions and Data Summary.

## Usage

```dart
import 'package:devaloop_data_list_view/data_list_view.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Data List View',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.blue.shade900),
        useMaterial3: true,
      ),
      home: Scaffold(
        body: Container(
          color: Colors.grey.shade100,
          child: Card(
            elevation: 0,
            color: Colors.white,
            margin: const EdgeInsets.all(15),
            child: Padding(
              padding: const EdgeInsets.symmetric(horizontal: 15),
              child: DataListView(
                title: 'Clinical Checkups',
                dataItems: [
                  DataItem(
                    title: 'Blood Complete',
                    subtitle: const Text('Blood check up complete'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'ACA IgG',
                    subtitle: const Text(
                        'Cardiolipin antibodies (ACA) are autoantibodies produced by the body\'s immune system which mistakenly attacks cardiolipin, a substance found in the outer layer of cells (cell membranes) and platelets. These autoantibodies can affect the body\'s ability to regulate blood clotting.'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'ACA IgM',
                    subtitle: const Text(
                        'Cardiolipin antibodies (ACA) are autoantibodies produced by the body\'s immune system which mistakenly attacks cardiolipin, a substance found in the outer layer of cells (cell membranes) and platelets. These autoantibodies can affect the body\'s ability to regulate blood clotting.'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'Anti-Beta2-Glikoprotein 1',
                    subtitle: const Text(
                        'Anti-β2 Glycoprotein 1 is the main protein that binds to phospholipid and cardiolipin antibodies.'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'APTT',
                    subtitle: const Text(
                        'Activated partial thromboplastin time (APTT) is a laboratory test to assess the activity of intrinsic and common pathway coagulation factors, namely factor factor IX (Christmas factor), factor VIII (antihemophilic factor, AHF), factor X (Stuart factor), factor V (proaccelerin), factor II (prothrombin) and factor I (fibrinogen). This test is for monitoring heparin therapy or the presence of circulating anticoagulant. The APTT is prolonged due to a deficiency of intrinsic coagulation factors and if the level is <> 7 seconds from the normal value, then the examination results are considered abnormal.'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'AT III',
                    subtitle: const Text(
                        'Antithrombin III (ATIII) is a vitamin K-independent protease that inhibits coagulation by neutralizing the activity of thrombin enzymes (factors IIa, IXa, Xa). Antithrombin III activity is markedly potentiated by heparin, the primary mechanism by which heparin and low molecular weight heparins produce anticoagulation.'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'CD 34',
                    subtitle: const Text(
                        'CD34 is a protein found in living creatures such as humans, mice, rats and other species. Even though it is found in the body, such as in the bone marrow, blood vessel lining and certain cells, CD34 is also found in several types of tumor cells.'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'Colestrol',
                    subtitle: const Text('Blood checkup complete'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'Colestrol',
                    subtitle: const Text('Blood checkup complete'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'Colestrol',
                    subtitle: const Text('Blood checkup complete'),
                    onTap: () {},
                  ),
                ],
                actions: [
                  ActionButton(
                    onPressed: () {},
                    label: 'Add',
                    icon: const Icon(Icons.add),
                    showForward: true,
                  ),
                  ActionButton(
                    onPressed: () {},
                    label: 'Search',
                    icon: const Icon(Icons.search),
                    showForward: true,
                  ),
                  ActionButton(
                    onPressed: () {},
                    label: 'Refresh',
                    icon: const Icon(Icons.refresh),
                    showForward: true,
                  ),
                  ActionButton(
                    onPressed: () {},
                    label: 'Categories',
                    icon: const Icon(Icons.category),
                  ),
                  ActionButton(
                    onPressed: () {},
                    label: 'Type',
                    icon: const Icon(Icons.type_specimen),
                  ),
                ],
                summaries: const [
                  Summary(
                    value: '150',
                    label: 'Total',
                  ),
                  Summary(
                    value: '100',
                    label: 'Added',
                  ),
                  Summary(
                    value: '50',
                    label: 'Updated',
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```


CHANGELOG.md
## 0.0.1

* Data List View: Data Items List, Data Actions and Data Summary.


lib/data_list_view.dart
library devaloop_data_list_view;

import 'package:flutter/material.dart';

class DataListView extends StatelessWidget {
  const DataListView({
    super.key,
    required this.title,
    required this.dataItems,
    this.actions,
    this.summaries,
    this.actionColor,
  });

  final String title;
  final List<ActionButton>? actions;
  final List<Summary>? summaries;
  final List<DataItem> dataItems;
  final Color? actionColor;

  @override
  Widget build(BuildContext context) {
    var orienstation = MediaQuery.of(context).orientation;
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        Row(
          children: [
            SizedBox(
              height: 24 + 15,
              child: Align(
                alignment: Alignment.centerLeft,
                child: Text(
                  title,
                  style: const TextStyle(
                    fontSize: 14,
                  ),
                ),
              ),
            ),
            if (actions != null)
              if (actions!
                  .where((element) => element.showForward == true)
                  .isEmpty)
                const Expanded(
                  child: Text(''),
                ),
            if (actions != null)
              if (actions!
                  .where((element) => element.showForward == true)
                  .isNotEmpty)
                Expanded(
                  child: SingleChildScrollView(
                    scrollDirection: Axis.horizontal,
                    reverse: true,
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.end,
                      children: actions!
                          .where((element) => element.showForward == true)
                          .map(
                            (e) => InkWell(
                              child: Padding(
                                padding: const EdgeInsets.all(0),
                                child: Row(
                                  mainAxisSize: MainAxisSize.min,
                                  children: [
                                    if (e.icon != null)
                                      IconButton(
                                        onPressed: () => e.onPressed.call(),
                                        icon: e.icon!,
                                        color: actionColor ??
                                            Theme.of(context)
                                                .colorScheme
                                                .primary,
                                      ),
                                  ],
                                ),
                              ),
                              onTap: () => e.onPressed.call(),
                            ),
                          )
                          .toList(),
                    ),
                  ),
                ),
            if (actions != null)
              InkWell(
                onTap: () {
                  showDialog(
                    context: context,
                    builder: (context) {
                      return AlertDialog(
                        title: Row(
                          children: [
                            Text(title),
                            const Expanded(
                              child: Text(''),
                            ),
                            IconButton(
                              onPressed: () {
                                Navigator.of(context).pop();
                              },
                              icon: Icon(
                                Icons.close,
                                color: actionColor ??
                                    Theme.of(context).colorScheme.primary,
                              ),
                            ),
                          ],
                        ),
                        content: SizedBox(
                          width: 320,
                          child: SingleChildScrollView(
                            child: Column(
                              mainAxisSize: MainAxisSize.min,
                              children: [
                                ListView.builder(
                                  shrinkWrap: true,
                                  physics: const NeverScrollableScrollPhysics(),
                                  itemBuilder: (context, index) {
                                    return ListTile(
                                      onTap: () =>
                                          actions![index].onPressed.call(),
                                      leading: IconButton(
                                        color: actionColor ??
                                            Theme.of(context)
                                                .colorScheme
                                                .primary,
                                        icon: actions![index].icon!,
                                        onPressed: () =>
                                            actions![index].onPressed.call(),
                                      ),
                                      title: Text(
                                        actions![index].label,
                                        style: TextStyle(
                                          color: actionColor ??
                                              Theme.of(context)
                                                  .colorScheme
                                                  .primary,
                                        ),
                                      ),
                                    );
                                  },
                                  itemCount: actions!.length,
                                ),
                              ],
                            ),
                          ),
                        ),
                      );
                    },
                  );
                },
                child: Padding(
                  padding: const EdgeInsets.all(7.5),
                  child: Icon(
                    Icons.more_vert,
                    color: actionColor ?? Theme.of(context).colorScheme.primary,
                  ),
                ),
              ),
          ],
        ),
        if (orienstation == Orientation.portrait)
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.stretch,
              children: [
                if (summaries != null)
                  Container(
                    margin: const EdgeInsets.only(top: 7.5),
                    child: Wrap(
                      alignment: WrapAlignment.spaceEvenly,
                      spacing: 7.5,
                      children: summaries!
                          .map(
                            (e) => Column(
                              mainAxisAlignment: MainAxisAlignment.center,
                              children: [
                                Text(
                                  e.value,
                                  textAlign: TextAlign.center,
                                ),
                                Text(
                                  e.label,
                                  textAlign: TextAlign.center,
                                ),
                              ],
                            ),
                          )
                          .toList(),
                    ),
                  ),
                if (summaries != null)
                  Container(
                    margin: const EdgeInsets.only(left: 7.5, right: 7.5),
                    child: const Divider(),
                  ),
                Expanded(
                  child: SingleChildScrollView(
                    child: Column(
                      mainAxisSize: MainAxisSize.min,
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text('Show top ${dataItems.length}'),
                        Flexible(
                          child: ListView.separated(
                            physics: const NeverScrollableScrollPhysics(),
                            shrinkWrap: true,
                            itemBuilder: (context, index) {
                              return ListTile(
                                title: Text(dataItems[index].title),
                                subtitle: dataItems[index].subtitle,
                                trailing: dataItems[index].onTap == null
                                    ? null
                                    : InkWell(
                                        onDoubleTap: () =>
                                            dataItems[index].onTap!.call(),
                                        child: Icon(
                                          Icons.navigate_next,
                                          color: actionColor ??
                                              Theme.of(context)
                                                  .colorScheme
                                                  .primary,
                                        ),
                                      ),
                              );
                            },
                            separatorBuilder: (context, index) {
                              return const Divider();
                            },
                            itemCount: dataItems.length,
                          ),
                        ),
                      ],
                    ),
                  ),
                ),
              ],
            ),
          ),
        if (orienstation == Orientation.landscape)
          Expanded(
            child: Row(
              crossAxisAlignment: CrossAxisAlignment.stretch,
              children: [
                if (summaries != null)
                  Container(
                    margin: const EdgeInsets.all(0),
                    child: RotatedBox(
                      quarterTurns: 1,
                      child: Wrap(
                        alignment: WrapAlignment.spaceEvenly,
                        spacing: 7.5,
                        children: summaries!
                            .map(
                              (e) => RotatedBox(
                                quarterTurns: 2,
                                child: Column(
                                  mainAxisAlignment: MainAxisAlignment.center,
                                  children: [
                                    Text(
                                      e.value,
                                      textAlign: TextAlign.center,
                                    ),
                                    Text(
                                      e.label,
                                      textAlign: TextAlign.center,
                                    ),
                                  ],
                                ),
                              ),
                            )
                            .toList(),
                      ),
                    ),
                  ),
                if (summaries != null)
                  Container(
                    margin: const EdgeInsets.only(
                        left: 7.5, right: 7.5, bottom: 7.5),
                    child: const VerticalDivider(),
                  ),
                Expanded(
                  child: SingleChildScrollView(
                    child: Column(
                      mainAxisSize: MainAxisSize.min,
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text('Show top ${dataItems.length}'),
                        Flexible(
                          child: ListView.separated(
                            physics: const NeverScrollableScrollPhysics(),
                            shrinkWrap: true,
                            itemBuilder: (context, index) {
                              return ListTile(
                                title: Text(dataItems[index].title),
                                subtitle: dataItems[index].subtitle,
                                trailing: dataItems[index].onTap == null
                                    ? null
                                    : InkWell(
                                        onDoubleTap: () =>
                                            dataItems[index].onTap!.call(),
                                        child: const Icon(Icons.navigate_next),
                                      ),
                              );
                            },
                            separatorBuilder: (context, index) {
                              return const Divider();
                            },
                            itemCount: dataItems.length,
                          ),
                        ),
                      ],
                    ),
                  ),
                ),
              ],
            ),
          ),
      ],
    );
  }
}

class ActionButton {
  const ActionButton({
    required this.onPressed,
    required this.label,
    this.icon,
    this.showForward,
  });
  final Function() onPressed;
  final String label;
  final Widget? icon;
  final bool? showForward;
}

class Summary {
  const Summary({
    required this.value,
    required this.label,
  });
  final String value;
  final String label;
}

class DataItem {
  const DataItem({
    required this.title,
    required this.subtitle,
    this.onTap,
  });
  final String title;
  final Widget? subtitle;
  final Function()? onTap;
}


example/lib/main.dart
import 'package:devaloop_data_list_view/data_list_view.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Data List View',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.blue.shade900),
        useMaterial3: true,
      ),
      home: Scaffold(
        body: Container(
          color: Colors.grey.shade100,
          child: Card(
            elevation: 0,
            color: Colors.white,
            margin: const EdgeInsets.all(15),
            child: Padding(
              padding: const EdgeInsets.symmetric(horizontal: 15),
              child: DataListView(
                title: 'Clinical Checkups',
                dataItems: [
                  DataItem(
                    title: 'Blood Complete',
                    subtitle: const Text('Blood check up complete'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'ACA IgG',
                    subtitle: const Text(
                        'Cardiolipin antibodies (ACA) are autoantibodies produced by the body\'s immune system which mistakenly attacks cardiolipin, a substance found in the outer layer of cells (cell membranes) and platelets. These autoantibodies can affect the body\'s ability to regulate blood clotting.'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'ACA IgM',
                    subtitle: const Text(
                        'Cardiolipin antibodies (ACA) are autoantibodies produced by the body\'s immune system which mistakenly attacks cardiolipin, a substance found in the outer layer of cells (cell membranes) and platelets. These autoantibodies can affect the body\'s ability to regulate blood clotting.'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'Anti-Beta2-Glikoprotein 1',
                    subtitle: const Text(
                        'Anti-β2 Glycoprotein 1 is the main protein that binds to phospholipid and cardiolipin antibodies.'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'APTT',
                    subtitle: const Text(
                        'Activated partial thromboplastin time (APTT) is a laboratory test to assess the activity of intrinsic and common pathway coagulation factors, namely factor factor IX (Christmas factor), factor VIII (antihemophilic factor, AHF), factor X (Stuart factor), factor V (proaccelerin), factor II (prothrombin) and factor I (fibrinogen). This test is for monitoring heparin therapy or the presence of circulating anticoagulant. The APTT is prolonged due to a deficiency of intrinsic coagulation factors and if the level is <> 7 seconds from the normal value, then the examination results are considered abnormal.'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'AT III',
                    subtitle: const Text(
                        'Antithrombin III (ATIII) is a vitamin K-independent protease that inhibits coagulation by neutralizing the activity of thrombin enzymes (factors IIa, IXa, Xa). Antithrombin III activity is markedly potentiated by heparin, the primary mechanism by which heparin and low molecular weight heparins produce anticoagulation.'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'CD 34',
                    subtitle: const Text(
                        'CD34 is a protein found in living creatures such as humans, mice, rats and other species. Even though it is found in the body, such as in the bone marrow, blood vessel lining and certain cells, CD34 is also found in several types of tumor cells.'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'Colestrol',
                    subtitle: const Text('Blood checkup complete'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'Colestrol',
                    subtitle: const Text('Blood checkup complete'),
                    onTap: () {},
                  ),
                  DataItem(
                    title: 'Colestrol',
                    subtitle: const Text('Blood checkup complete'),
                    onTap: () {},
                  ),
                ],
                actions: [
                  ActionButton(
                    onPressed: () {},
                    label: 'Add',
                    icon: const Icon(Icons.add),
                    showForward: true,
                  ),
                  ActionButton(
                    onPressed: () {},
                    label: 'Search',
                    icon: const Icon(Icons.search),
                    showForward: true,
                  ),
                  ActionButton(
                    onPressed: () {},
                    label: 'Refresh',
                    icon: const Icon(Icons.refresh),
                    showForward: true,
                  ),
                  ActionButton(
                    onPressed: () {},
                    label: 'Categories',
                    icon: const Icon(Icons.category),
                  ),
                  ActionButton(
                    onPressed: () {},
                    label: 'Type',
                    icon: const Icon(Icons.type_specimen),
                  ),
                ],
                summaries: const [
                  Summary(
                    value: '150',
                    label: 'Total',
                  ),
                  Summary(
                    value: '100',
                    label: 'Added',
                  ),
                  Summary(
                    value: '50',
                    label: 'Updated',
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}
